### PR TITLE
type-safe denominations implemented

### PR DESCRIPTION
**Pull request by theodorebugnet**
Example currencies are included under `currencies/`; consumers can follow the same structure to easily add additional currencies.
Conversions are autogenerated on `MonetaryAmount` based on the units, providing a structure such as
```ts
const btcAmount = BTCAmount.from.BTC(1);
const satBig: Big = btcAmount.to.Satoshi();
console.log(btcAmount.str.BTC()); // "1"
```
which removes the need to verbosely subclass `MonetaryAmount` as was done in xopts.

Some documentation would probably be good.

### CODE REVIEW

File: src/monetary.ts [hunk 1]
```diff
@@ -1,85 +1,42 @@
...
-  toString(humanFriendly = false): string {
-    if (!humanFriendly) {
-      return this._amount.toString();
-    }
-    const amount = this.toBig(0);
-    // FIXME: would be better to format directly to locale string
-    // but after conversion to 0 decimals, amount should be small
-    // enough to fit in normal number
-    return parseFloat(amount.toString()).toLocaleString();
+  toString(unit?: U[keyof U]): string {
```
Comment (nud3l) [inline src/monetary.ts:33]: Would be nice to have a `toHuman` as well with a default decimal place. For example, could have a string with only 2,3,4,5,... decimals. Then you could do something like this in the UI without any helper functions:

```ts
dotAmount.toHuman() // default e.g.5 decimals
dotAmount.toHuman(8) // 8 decimals
```

Comment (theodorebugnet) [reply src/monetary.ts:33]: Good idea

**Review by nud3l — APPROVED** [review]

**Review by nud3l — APPROVED** [review]

File: src/exchangeRate.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  toRawBig(): Big {
+    return this.toBig({ baseUnit: 0, counterUnit: 0 } as UnitMap<
+      BaseUnit,
+      CounterUnit
+    >);
+  }
+
+  toString(
+    units?: Partial<UnitMap<BaseUnit, CounterUnit>>,
+    precision?: number
```
Comment (savudani8) [inline src/exchangeRate.ts:98]: An optional `roundingMode` parameter would be useful to pass as the second argument of `rate.round(...)`

File: src/monetary.ts [hunk 3]
```diff
@@ -1,85 +1,68 @@
...
-    readonly decimals: number = 18
-  ) {}
-}
-
-export class Tether extends ERC20 implements Currency {
-  constructor(address: string) {
-    super('Tether', address, 6);
+    readonly currency: C,
+    amount: BigSource,
+    unit: U[keyof U] = 0 as U[keyof U]
```
Comment (savudani8) [inline src/monetary.ts:43]: I'm not very familiar with this pattern and I'm wondering what the reasoning is behind having `U[keyof U]` in most places. Given `U extends UnitList`, won't `unit` always be `number`? Or since `unit` is passed to `Big` utility functions, I think it would at most be a `BigSource`?

Comment (theodorebugnet) [reply src/monetary.ts:43]: `U[keyof U] extends number`, yes, but it's restricted only to the list of numbers that are valid units. E.g. if
```
typeof U === {
  a: 1,
  b: 2
}
```
Then `typeof U[keyof U] === 1 | 2`.

In this way, type safety is achieved: you can convert e.g. a Bitcoin to `BTCUnit.BTC` (= 8) or `BTCUnit.Satoshi` (= 0) but you can't just pass an arbitrary `number` to attempt to convert to e.g. 7 decimal places: that's a type error, because `7` does not match `BTCUnit[keyof BTCUnit`.

File: src/monetary.ts [hunk 4]
```diff
@@ -113,119 +96,42 @@ export class MonetaryAmount<C extends Currency> {
...
-    return this.toBig(ETHUnit.GWei);
-  }
-
-  toEther(): Big {
-    return this.toBig(ETHUnit.Ether);
-  }
+    return new Cls(this.currency, amount, unit);
+  }
+
+  to: toConversions<U> = (() =>
```
Comment (savudani8) [inline src/monetary.ts:108]: I haven't seen this kind of syntax before, can you please point me to a reference?

File: test/exchangeRate.test.ts [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+import {
+  Bitcoin,
+  BTCAmount,
+  BTCUnit,
+  ETHAmount,
+  Ethereum,
+  ETHUnit,
+} from "../src/currencies";
+import { ExchangeRate } from "../src/exchangeRate";
+import * as fc from "fast-check";
```
Comment (savudani8) [inline test/exchangeRate.test.ts:12]: Is `fast-check` more expressive than Mocha?

File: src/monetary.ts (hunk 4 above)
Comment (theodorebugnet) [reply src/monetary.ts:108]: Which part are you referring to? This defines a property `to` on the class, of type `toConversions` which is defined above (line 15). The value of the property is set to the expression, which is an immediately-invoked function expression (IIFE) - an anonymous function is defined `() => {}`, and then this expression - in brackets `(() => {})` - is immediately called - `(() => {})()` - thus the entire expression evaluates to the return type of the function. In this case, the function iterates over the `entries` in the currency's UnitList, and for every `[key, value]` pair it returns a `[key, () => { /* convert to Big for that unit in the UnitList */ }]` pair, which `Object.fromEntries()` then assembles into an object using the keys and values.

Thus e.g.
```
{
  BTC: 8,
  Satoshi: 0,
}```
is transformed into
```
{
  BTC: () => this.toBig(8),
  Satoshi: () => this.toBig(0),
}
```
And this object is assigned to the `to` property, allowing one to use e.g. `btcAmount.to.BTC()` as a convenient shortcut to invoke `btcAmount.toBig(BTCUnits.BTC)`.

File: test/exchangeRate.test.ts (hunk 5 above)
Comment (theodorebugnet) [reply test/exchangeRate.test.ts:12]: `fast-check` uses property-based testing, where you define values to test over and assertions for those values. E.g. "for every Big(), ensure that `new BTCAmount(big).toString === big.toString()` holds".

To be honest the only reason I'm using it is because the original xopts tests used it, and when I rewrote the logic I didn't consider changing it. I think it's rather neat, but I haven't really spend much time comparing it to other frameworks.

We still use `mocha` as the test runner, by the way.

File: test/monetary.test.ts [hunk 6]
```diff
@@ -2,402 +2,270 @@ import Big, { BigSource } from "big.js";
...
-      it("should use currency number of decimals by default", () => {
-        const rawAmount = new Big(10);
-        const amount = new DummyAmount(rawAmount);
-        expect(amount.toBig().toString()).to.eq(rawAmount.toString());
-      });
-
-      it("should work with any number of decimals", () => {
+  describe("arithmetic", () => {
+    describe("eq", () => {
+      it("should equal when amounts are equal", () => {
```
Comment (savudani8) [inline test/monetary.test.ts:141]: Could add a similar test that creates `DummyAmount` of same value but from different denominations

File: src/monetary.ts (hunk 4 above)
Comment (savudani8) [reply src/monetary.ts:108]: Oh wow this is very nice. I asked the question because I got confused seeing the colon at the end of the class,but this additional explanation is very useful!

File: test/exchangeRate.test.ts (hunk 5 above)
Comment (savudani8) [reply test/exchangeRate.test.ts:12]: Didn't know such a thing existed, sounds great. I guess it does some sort of static analysis on the type? Maybe we could apply it in the lib too.

> We still use mocha as the test runner, by the way.

I'm probably confusing `mocha`, `chai` and `fast-check`. So `mocha` is the test runner, `chai` is as assertion utility, and `fast-check` is more similar to `chai` rather than `mocha`

**Review by savudani8 — APPROVED** [review]

File: test/exchangeRate.test.ts (hunk 5 above)
Comment (theodorebugnet) [reply test/exchangeRate.test.ts:12]: > I guess it does some sort of static analysis on the type

I haven't looked into this too deeply, but I believe it's largely randomised (like some kind of pseudo-fuzzing), plus some clever tricks to suss out edge cases as well as find minimal reproducing cases on failure. I didn't know about it either before copying the xopts tests over, but this is an interesting read: https://github.com/dubzzz/fast-check/blob/main/documentation/HowItWorks.md

> `mocha`, `chai` and `fast-check`

Pretty much. I suppose fast-check can be thought of as a test case generator that tests a range of values and tries to find edge cases for you, while chai gives the actual asserts that accept/reject each individual test case.
